### DISCUSSION TITLE
Thoughts on the console.

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#87] [General]
Given some projects I'd like to do, unlinking the console from DtCyber, into a four or so packet type raw socket pair (A out, B out, dot mode out, and operator input).

This opens the door to a plethora of console possibilities, including headless, text-mode console (even ncurses can keep up, I'd think), using remote consoles, without HDMI or VNC, with or without graphics.

With this there should be options to wait or not for an established connection, and what to do if not (script turning off debug, log,14 on the cybis packs, and the cs alert in nam, whatever you need to do on yours for boot)...

There should also be an option for AF_UNIX (iirc) type localhost-only sockets for a non-networked host.

The default console itself would be separately compiled, and independent, usable for local and remote.  Maybe along with a ncurses version.  ($DISPLAY undefined or invalid, or command-line override, use ncurses, maybe even a screen-like detached mode, otherwise X11 / Windows, or something to that effect).

If I had the time on top of what I am doing, this is the direction I'd like to go..  for the Pi stuff, I am coming to the conclusion that headless is the way to go.  Playing with zeroes.  May have to do it anyway to get things booted on a Pi Zero W 1 or 2 without any form of interaction, and requiring nothing but the board and the ability to get to at least a serial connection, with the valid assumption that serial and ssh might be the only means of comms with raspios / raspian / fbsd with the axiom that serial-only is a distinct possibility, with all networking unbridged intranet (that is a different topic, involving local caching, (--keep-distfiles ???)), an even more distinct possibility.

**Comment by kej715** [comment]
Some things to consider:

   - Paul Koning has implemented this feature previously in his private
   fork of DtCyber, and Cyber1.org uses it. Based upon comments in Paul's
   version of console.c, the feature seems very tricky to implement well.
   Instead of re-inventing Paul's wheel, it might be better to build upon it.
   - Phil Claridge has also been investigating this feature. Perhaps
   consider a collaboration?
   - The current console automation feature has become quite important. Any
   changes in the console area should avoid breaking the automation feature.
   - It might be worthwhile to consider leveraging the 2-port mux design.
   Cyber 800 series machines, including the 865, 875, and all 180's had a
   2-port mux, and NOS could be configured to use it to support its console
   instead of requiring a CC545. A CDC Viking 721 terminal could be connected
   to the 2-port mux and serve as the NOS system console. The Viking 721 was
   an async serial terminal. It substituted quite nicely for a CC545. NOS/VE
   did not use the CC545 as a console. It used only the 2-port mux interface.
   If a Viking 721 emulator was implemented (e.g., in Javascript as an
   addition to the existing aterm, pterm, tn3270, and tek4014 emulators), it
   might not be necessary to modify the console.c module to achieve what's
   desired.

cheers,
Kevin

**Reply by wschaub** [reply]
From the reading I've been doing on NOS/VE and the 180 in general it looks like the TPMUX style console is supported by a PP program called SCI (system console interface) which will listen on channel 10 and accept old style console commands/data. I think this was brought up by CIP early on if it determined it was needed.  SCI is shipped with NOS/VE sources it seems so we have a copy but I'm unsure how easy this would be to implement.  (getting this right would be another notch towards a working 180 simulation though, I'm just unsure that CIP would ever attempt to launch something like SCI on a system with a SCR register vs a full 180 style maintenance register.)  The SCI source listing I've read also looks like it uploads a special terminal resident into the 721 that is to act as a console.

**Comment by kej715** [comment]
I don't know for sure whether SCI can be started on an 865/875 to drive the
console, but it seems likely that it is possible. The ROI for designing the
2-port mux into the 865/875 purely for RDF seems low.

**Reply by wschaub** [reply]
I know I haven't read any documentation that suggests using a 721 as a console on an 865 or 875 is an option it would be nice to know for sure.   

It's also interesting to note that on a 180 extra function codes are supported by the TPMUX for things like getting the date/time off a clock chip embedded in the IOU I believe NOS can use this to bypass asking for date/time. and I think there's also a way to query the TPMUX to see if there's a real console controller inside the IOU for systems that support it as an option like on an 830.  

Aside from RDF I believe RMM or whatever it might be called under MSL15X would be another use for the TPMUX remotely managed offline CE diagnostics.  (another interesting thing to note BTW is that the SMM4 source has source code that can be built with the CROSS tools for the 2550 to load into an NPU to turn it into a 1 port MUX for RMM to use, if anyone was ever crazy enough to try building a real NPU emulator that runs real instructions it would be a probably the simplest possible test code to use to try and run on one)

**Comment by philclaridge** [comment]
Kevin,
I am indeed working my way through Paul’s fork of console.c as a potential basis for a very high quality console emulation and in time integrating with the existing websocket infrastructure to remote other terminal types. I am impressed that Paul undertook the full emulation of the X-Y tube to get the character distortion.

Personally I do want all the look and feel of the real Cyber console, and I am not sure that the CDC Viking 721 does all of the character sizing and will give the same high refresh feel hat I am looking for on a serial interface (e.g. with superimposed characters during boot). Not to say that the Viking 721 is not worth progressing.

Also see https://github.com/kej715/DtCyber/blob/main/BUILDING.README.txt walk though running NOS on different Raspberry Pi. Runs nicely on a Pi 2 Zero W.

Happy to chat to others about this. 
Phil

**Reply by KC5VDJ** [reply]
I have to check this out.  As for the 721, I have been toying with a terminal emulator, but it would be a large project, I have a ton of terminal manuals, and if I do it, it will have weirdos (6530 as well).  The last time I did serial code was some time back, and the real stuff was in masm/c (isr and text-mode video in x86 assembly).  

But, I might be convinced to.  Seeing what I recently saw when looking at the price of one with better coverage than others ($650?!?!), combined my wife and I's situation, writing one looks tempting, and maybe pay some bills..  I'll need an 8k monitor to do the tek terms properly 😁.  43 inches is perfect for 4k monitors, but 2160p won't fit a tek term on the screen in full emulation.  At least 4014 and 4016, and possibly the 15 (APL) up to full 4096, with realistic cursor, would be the initial in that line.  Panning locally, in addition to panning the tek internally, would be a, what's the word? A clusterf...  An 8k monitor is mandatory, IMHO, for full tek emulation.  Blame Tektronics, they made it 4096x4096 on a terminal without that Y physically.

Hmmmmmmmmmmmmmmm....  I'd start with CDC Viking 721, and VT-525 (a major undertaking itself), all the supported vt525 term emulations, then maybe 3270 and 6530, tek, ad nauseum...  this could work.  I could keep the 721 maybe free (aka: shareware, beerware, budware).

I'd be tempted to take some of the money, put together a terminal museum with the terms all connected to original running oses on pis stuffed inside them running emulators, the irony!!  I've been kinda hooked on this stuff since simh was on decwrl in the 90s, and booting interactive v6 and v7 (haven't touched rsts/e or rsx-11m since high school, 11/44).

**Comment by KC5VDJ** [comment]
OYFG... Dtcyber struggling to maintain 40-45%.  Xorg and vncserver averaging 50-70%...  55916k resident, 92976k virtual...

Pi-DI will happen.... But a text-mode console might be beneficial...  Well, I got to see what happens in the memtest for the first time...for an agonizingly long time....  still deadstarting the same install full images I made recently....  might watch a movie and see how long deadstart takes...

-O3 -m*=native

Thought about -fomit-frame-pointers, but no.  Improvement there would be minimal.

**Reply by KC5VDJ** [reply]
It runs on the bcm2835...no errors so far...

"Deadstart status.

Loading 170"

ps -o start 1778
 STARTED
21:32:31

UMMM...  It's like 22:21...

I'm having serious second thoughts about node install full with this running a distant second fiddle to the console via VNC on this, but I now almost have a perverse interest in how many months it would take...

I could try the raspios lite, but top doesn't show much help..  at best a small fraction of 27-33% would be regained (average xorg usage using raspios full).

**Comment by kej715** [comment]
Yup, after reviewing OSB, DIO, and SET, it's clear that CTI decides whether
the system console is a CC545 or a Viking 721, and it provides this info to
OSB, DIO, and SET. If CTI indicates that the console is a 721, DIO will
arrange to load the driver for it, SCD. It looks like an 865/875 -could-
have a 721 as its console, but the version of CTI we're currently using
doesn't seem to provide a way to select that as an option. Or, as William
has suggested, perhaps CTI would offer this option only on a "true" Cyber
180.

As an anecdote ... UMass replaced its Cyber 175 with a 2-CPU/64 Mbyte Cyber
870 in 1988. The 870 was a high-end dual state machine. As it was a 2-CPU
machine, its physical shape was a "T". We chose to use a Viking 721 as a
NOS system console instead of using a CC545. The 721 behaved very much like
a CC545. IIRC, it even supported "graphical" games like BAT, WRM, ADC, etc.
The 721 was a top-shelf async terminal for that era. It had an embedded,
programmable Z-80 processor, and NOS (or perhaps CTI) leveraged that
feature to download and run software that enabled it to emulate a CC545
pretty well.

The Cyber 870 at UMass also ran NOS/VE. NOS/VE never supported the CC545 as
its system console. It used either a Viking 721 or a Z-80 based PC as its
console.

cheers,
Kevin

**Reply by wschaub** [reply]
If you look at the source code for SCI you will find that there is a custom resident uploaded to the 721 to do just as you described.

There's also some differences between using a 721 this way and a real CC545. I would have to dig up the NOS software release bulletin that details it, but essentially a bunch of NOS displays were changed to accommodate these differences.  this included making the channel status display its own screen. (I think it had to do with how many lines can be displayed at a time but like I said I need to dig up the SRB to cite it) 

CTI is probably going to be doing one of either of:
1) check for maint channel or SCR
  if maint channel find out about the console either by talking to TPMUX directly for status or finding info in the maint channel

2) talking direct to the tpmux to find if we have a real console or not and launching the correct driver PP. 

Considering that in order to deadstart a CIP tape even on a 180 it has to figure this all out pretty early to give the initial operator display the logic would have to be fairly simple. 

When looking into various CIP/CTI versions I found that the 180 style CIP/CTI seemed to be simpler it shares a lot of the same CTI components but there's a lot of programs that apply to 170 specific stuff that are simply not there anymore because of just how rich the new 180 style maint channel capabilities are to both determine the exact mainframe config and also to give management commands to the various hardware subsystems in the machine that would otherwise take a lot more work on a more traditional 170 setup. 

I want to get a PP disassembler so I can tear apart CIP L716B this version may have a different driver than SCI but the effect would still be essentially the same something needs to listen on channel 10 and push stuff out the TPMUX.

I'm kind of interested in taking the very early circa 1979 CTI sources  and adding in the missing bits to implement the CTI/OS interface.  but I lack two manuals necessary to do so.  There's probably enough OS source code to make it possible to just re-write the APP program that passes the configuration information to the operating system but it's a big task for not a whole lot of gain (other than re-discovering how the CTI/OS interface really works and maybe helping with bringup on new emulators since you can choose how picky CTI is during the early stages and report back what you want if you can have your own custom CTI) 

Unfortunately the very early 1979 CTI does not implement the APP PP program at all. it's still very useful to use as a reference though regardless. but what I really want to do is dissassemble L716B (the version we use on the ds tape) and compare each component to the ones that exist in 1979. and try to annotate the stuff nobody has source for as best as I can.  

I really would love to see a fully working 721 that can run terminal residents. it would give us a full function terminal that's historically interesting all on its own.  I'm currently aware of an attempt to make a 721 terminal work inside MAME but I'm not sure how much of that is finished or if it's still an active project.

**Reply by KC5VDJ** [reply]
Something like sixel graphics?

I had graphics in quotes there, then again, the voice server at the sprint Lenexa ossc had a vt-330 that did a reatime fft waveform display when entering voice prompts into the mickeyvax 2 that was the admin processor from a cassette deck for a volt-maintech i386 based system running interactive unix which was at all the switches.  Sixel was quite usable.

(Wondering who moved into that computer room after sprint?  The whole bldg screams "go away" last time I saw it, FEMA maintaining the SOC that was co-located? FEMA didn't need that kind of physical security when sprint had the ldd ossc there).

**Reply by wschaub** [reply]
721 graphics aren't baked in to a ROM like sixel on a DEC terminal. (I mean it might have something native I'm not 100% up to speed with it) it's got a fully programmable Z80 on board and you can upload a resident to implement any graphics protocol you can dream up along with changing just about any terminal behavior you want. I believe there was even add-on hardware to give it a disk drive and run CP/M right on the terminal itself.  

I think later versions of the cyber console replaced the 721 with a zenith data systems PC running special software on top of DOS I haven't studied that era of the cyber too hard though, but I did read some materials about the Cyber 990 that talked a lot about it. it would be nice to have a backup of the software that runs on that PC.

**Reply by KC5VDJ** [reply]
Interesting.  This should be fun.  Kinda like a vt-180 then (a strange beast indeed).

I've done "searches" for a few months since reading about it in manuals, and come to zilch on the PC console software. Thanks to Google, etc, computer searches have gone to sh..  whatever happened to booleans and nearbys? Would it have been serial or using an interface board like with IBM?  The heath-zenith thing makes sense though, DOD kept CDC and Heath/Zenith floating (as Heathkit was on it's deathbed, sadly), and Desktop 1-4 inter-service contracts were heavy on the heath/zds.  Anyone ever have to play with that polled (no irq) serial on the z-248?  Ugh...  (For connecting to cybers at that), freaking loved those alps keyswitches tho, kicked a.. compared to these cherry blues.

Anyhow, If my guess is right, there may be copies, but these aren't the days of base bbs systems to download it from, and they may be in boxes on military bases to never be seen again.  ("We have top men on it...top men").

Sadly, it is easier to find pc6530 online, as obscure as that is.

**Reply by KC5VDJ** [reply]
Still, the PC console thing was a much better, and cheaper, solution than DEC using bastardized PRO 380s for consoles.  But, theirs was also equivalent to an 11/70 as far as the CPU went.  

Z80 is a little easier to deal with.  That, and well, Danny Cohen raised some interesting points in "on holy wars...", IIRC, concerning the PDP 11s. (https://gwern.net/doc/cs/algorithm/1981-cohen.pdf or https://www.rfc-editor.org/ien/ien137.txt).

**Comment by KC5VDJ** [comment]
Just out of curiosity, how long would a real 875 take to deadstart?  Or even get to "loading system library"?

Two hours in on a Pi Zero W via VNC, and "Loading ROYALTY"

No apparent errors yet.  Not really expecting any at this point.  I expect that the only problems are the usual for the build (dealing with the cybis pack message and cs alert, and http and rexec being unconfigured in the printed dayfiles (http maybe sometime, but I would rather not on rexec).

edit: just hit system lib....after 2:23 hours elapsed.  Wonder if it'll be quiescent by the time I wake up in the morning?

![IMG_20230905_234558389_HDR.jpg](https://github.com/kej715/DtCyber/assets/136194499/4653c4a4-98e4-42dc-b653-fa2c29ffde48)

This just gives me motivation to do Pi-DI....  looking at some bsd-type licensed code that could be derived from simh, at least the core, 68000, '010, '020, '030 '040.  All I'll need is 68k, and iirc, the '030 (or was it the '040 in the third (second "model") option for a DI controller board.  (Model 1, 32k; model 1, 128k; model 2, 4m to 16m in 4m increments)

**Reply by KC5VDJ** [reply]
Actually, generating the system libs didn't take too long....  Interesting though, since the console went nuts when it switched to the AB. display (as if it wasn't bad already), I minimized the console, and DtCyber went from the mid 30's to 83-85% CPU...

From what I can see, dtcyber was averaging low-mid 40's, with the console up during the "loading" phase, peaking for brief periods as high as 52.9%, and as low as upper 30's.

At all times all but maybe five or so percent of the rest was xorg and the vnc server.  Almost two thirds of the CPU, at times slightly over.

The only other active "applications" were three open terminal tabs, one kicked off node start and has that, next one is top with idles hidden, and the third, just a GP idle shell for whatever, and whatever wasn't happening, except to check start time, including wasting time doing the same thing using ps, ps, grep, and awk to give the pid of DtCyber for the ps -o start, inline, in a nice chain of pipes.  Things to do when bored.

**Reply by KC5VDJ** [reply]
Status update.

8am.  VNC seems to have timed out the windows connection, dtcyber is apparently running (80's% CPU), but I suspect that it may be timing out a reconnect to VNC, as I can't reconnect (it is running as a server on the pi0w1, and shows activity when trying to reconnect, but always times out on connection attempt).

All I know is that node start was called as root, thus ports should have opened, although there was an issue that 2550 couldn't be opened on start.  

I still can't telnet localhost on the Pi.

**Comment by wschaub** [comment]
Deadstart is mostly I/O bound so on a real thing I think the main thing holding back deadstart speed would be tape or disk speed. 2 hours I would be very surprised if anything real was that slow to fully dead start. (I've never brought up a real cyber but it's hard to believe the real thing would ever take 2 hours that would be unacceptable performance)

**Comment by kej715** [comment]
I brought up multiple models of real Cybers many times, and William is absolutely correct that the time needed to complete a deadstart was dictated by the speed of the deadstart device (tape or disk). Deadstarting a real Cyber 875 from tape probably wouldn’t have taken more than two or three minutes, and maybe less.On Sep 6, 2023, at 11:46 AM, William Schaub ***@***.***> wrote:﻿
Deadstart is mostly I/O bound so on a real thing I think the main thing holding back deadstart speed would be tape or disk speed. 2 hours I would be very surprised if anything real was that slow to fully dead start. (I've never brought up a real cyber but it's hard to believe the real thing would ever take 2 hours that would be unacceptable performance)

**Comment by kej715** [comment]
Yes, the 721 could be attached to a floppy drive and run CP/M, like the IST-III terminal that it effectively replaced. William is also correct about the Zenith PC’s serving as a console. Every Cyber 930 came with one. Probably every 960 too.  Not sure about the 990 or Cyber 2000. Thw 990 was a dual state machine, so presumably could have a CC545 as NOS console. IIRC, the Cyber 2000 and 2000V were VE-only, like the 930, so probably came with a Zenith PC also.  In fact, IIRC, the Zenith PC loaded microcode into the 930 and 2000, and perhaps other 180 models too. The 2000V had vector instructions (hence the “V”), like the 990 in 180 mode.On Sep 6, 2023, at 12:34 PM, William Schaub ***@***.***> wrote:﻿
721 graphics aren't baked in to a ROM like sixel on a DEC terminal. (I mean it might have something native I'm not 100% up to speed with it) it's got a fully programmable Z80 on board and you can upload a resident to implement any graphics protocol you can dream up along with changing just about any terminal behavior you want. I believe there was even add-on hardware to give it a disk drive and run CP/M right on the terminal itself.
I think later versions of the cyber console replaced the 721 with a zenith data systems PC running special software on top of DOS I haven't studied that era of the cyber too hard though, but I did read some materials about the Cyber 990 that talked a lot about it. it would be nice to have a backup of the software that runs on that PC.

**Reply by KC5VDJ** [reply]
I would assume the microcode is with the software wherever that is?

**Comment by kej715** [comment]
The 721 had an async serial interface. When used as a console, it probably ran at 19.2Kbps. I’m pretty sure that the Zenith PC also connected to the mainframe using an async serial interface (e.g., RS-232).As an aside, the Cyber 930 was a wonderful NOS/VE machine. It was nicknamed “Darth Vader’s Washing Machine” because it was black and about the size of a clothes washer. The usual disk and tape subsystem coupled with it had a matching black cabinet, so looked like Darth’s clothes dryer. The Zenith PC console was usually perched atop the CPU cabinet. The 930 also had a CDCNet device called an ICA (Integrated Communication Adapter) which was basically an MDI/ESCI combination, providing the machine with an integrated Ethernet connection and networking capability. On Sep 6, 2023, at 1:51 PM, KC5VDJ ***@***.***> wrote:﻿
Interesting.  This should be fun.  Kinda like a vt-180 then (a strange beast indeed).
I've done "searches", and come to zilch on the PC console software.  Would it have been serial or using an interface board like with IBM?  The heath-zenith thing makes sense though, DOD kept CDC floating, and desktop 1, desktop 2, desktop 3, and desktop 4 inter-service contracts were heavy on the heath/zds.  Anyone ever have to play with that polled (no irq) serial on the z-248?  Ugh...  (For connecting to cybers at that).

**Reply by KC5VDJ** [reply]
Well, as I was tarballing my home dir before upgrading to the latest raspios, I was also doing an end to end sequential read of the sdhc, and it averaged 19-20 MB/s....  Random has to be worse.  Add in the extreme overhead of a modern xorg, wm, and vnc......all on a single core and sdhc...

Fbsd might do it a little faster, but in it's current situation, this dtcyber is seriously io bound, and lack of a monitor is taking a third of the cycles. (vnc).  I was expecting this on the 0W, but not so bad...

Telnet still refusing, and vnc tries but can't connect.  DtCyber still over 80% CPU though without transmitting the desktop.   Wanna pull the plug, but holding out.

Hmm.  Just noticed that 2550 is listening, despite the earlier message that it wasn't.  Ssh works.

**Reply by wschaub** [reply]
Even with the 930 being NOS/VE only it's got to be the smallest cyber built would be nice if one of those got preserved.

**Reply by KC5VDJ** [reply]
Sounds like it.  Kinda reminds me of the way I used to use a sun ue450.  Monitor, kb, and mousepad yes, that freaking metal thing that the mouse wouldn't work without (still better than Engelbart's in the mother of all demos) all on top.  Someone should shoot the designer of the sun model 5 keyboard, ever accidentally turn off the server just by picking up a keyboard?  I have.  Freaking moron put a power key in the upper corner.  Repeated a couple of more times at home on a couple of Ultra AXi systems (UltraSPARC IIi's @300 mhz, pci-x, atx form factor in stealth pc cases, if anyone needs a working obp image for one, those batteries are long dead now, I found it the other day in my archive (also the netboot package for hp 700, envizex (sp?), and entria x-terms)..

That black looks slick.  Makes one wonder if he was involved with the eta-10, or influenced whomever. Slick black...

**Comment by philclaridge** [comment]
As there are different console futures I have labeled some ...

Option A: Serial console. Is the serial protocol/decode documented anywhere for any of the terminal based console solutions? If not could we spin up a host system (I don't have the CDC expertise to reconfigure) and see if we can capture some serial output while the machine is booting?

Option B: Re-integrate Paul Koning’s code (perhaps allowing for X or serial connection) I am happy I mostly understand how Paul Koning’s console.c code works and think it would be one basis for a remote console, though the remote display side of Paul’s code should be rewritten. 

Option C: (New in this thread) Remote all of the console function. However the radical and fun solution I am considering that would is if I could bridge the raw CDC channel to which console.c normally interfaces with to a local TCP connection, and then write an intermediate JS/TS server that exclusively uses this local connection, and can then fan out to multiple consoles using websockets. This JS/TS server would de-duplicate the channel data (c.f. Paul Koning’s solution) to reduce remote bandwidth, and could even contain a virtual character screen buffer to drive an existing remote character based terminal emulation. I am sure some of the existing JS infrastructure could help as foundation. The same JS/TS server can act as an injection point for any keystroke automation.

**Reply by wschaub** [reply]
I would like to be able to still run dtcyber standalone without needing a separate nodejs component. it would be nice to be able to select whatever console scheme is being used inside cyber.ini it already seems like console.c has a bit of a separation from window_x11.c and window_win32.c  whatever solution we go with for enhanced consoles should still support the DSD typein stuff we've added for automation obviously.  

(I wouldn't mind having, even if it was limited functionality a ncurses console to replace window_x11.c for stuff that I'm running on a headless system like the Pi. I could eliminate having to run a VNC server and I could just use something like screen or tmux instead) 

one thing that would be nice for any new graphical console implementation really would be a nice capable cross platform graphics library like SDL2 or whatever seems decent.   

I have tried to run paul's dtcyber in the past but have run into strange behavior with his console. I'm not sure if it was related to my particular graphics stack or what (ubuntu X11) but I've never been successful with it. I may need to retry now that I'm several major ubuntu versions ahead from the last time I attempted it. 


To answer your question about finding out what is sent to the terminal over the serial port  I think the source code to SCI + the source for the special terminal resident for the 721 could provide the way forward for knowing what is spoken to the 721 style console.  I have sources for those and most of my comments on the serial based console are based on what I've read so far skimming through it.

**Reply by KC5VDJ** [reply]
A  Y cable capture could be useful regardless.

What I am thinking is closer to option c, keeping the current or modified console for those of us that like the local console experience, and whatever for text mode, and the scripting possibilities...

It also fits into the whole Pi-DI concept making dtcyber what it is supposed to emulate, a distributed computing paradigm, making things like Pi-DI possible.  Using sockets (inet and af_unix) instead of cables.  The performance hit would depend largely on os tcp/ip stack performance, and will probably only get better with time and Moore's "Law"., Despite Moore's "Law" lagging badly these days.  Amdahl's "Law" is holding up better.

**Comment by kej715** [comment]
Yes, the 930 was the smallest Cyber built. It was very cute and capable.
The 810, 815, 825, and 830 were also small, air-cooled machines, and they
were all dual-state. They weren't much bigger than a VAX-11/780.

**Reply by philclaridge** [reply]
On topic of small machines, here is my smallest emulation to date. Need to adjust some character sizing / spacing though. 
![IMG_0767](https://github.com/kej715/DtCyber/assets/126126/e7d844c9-7d88-4484-bad7-865bf919c9b5)
![IMG_0768](https://github.com/kej715/DtCyber/assets/126126/25af7cdb-6827-46aa-8f4c-ad236e405085)

**Reply by wschaub** [reply]
The 830 is on my radar for a good candidate for a 180 emulation. there's a very very detailed principles of operation manual for it that goes deep down to the microcode level of both the IOU and the CPU. there's no microcode listing for the CPU in there but there is essentially for the IOU and there is discussion of the CPU microengine.  http://bitsavers.trailing-edge.com/pdf/cdc/cyber/cyber_180/Cyber_810_830/60469460A_Cyber_180_810_830_Central_Computer_Theory_of_Oper_Jun84.pdf

I don't expect an emulator to appear any time soon since it's non trivial but the 830 PDF seems to be a really good hardware specific compliment to the model independent general design spec as far as filling in a lot of important blanks.  The CTI/OS specification and the 180 configuration handbook  + the DFT/OS interface specification would fill in even more. if they could be found.  

I'm actually kind of curious about faking enough through the maintenance register to "fake out" a CIP to successfully load NOS and think it's operating under EI control without a real 180 CPU being under the hood. basically pretend we have the microcode and EI already running and we're just re-deadstarting in a 170 state only situation.  I kind of get the feeling that may be what was going on with the experimental 840A model in init.c but I don't know for sure.  There's a lot of stuff in the IOU side that's unique to the 180 series that seems fairly easy to implement on current dtcyber at least enough to experiment with a bit.  

There's definitely a bit of a chicken and egg problem though since I think all the available tools to build 180 virtual state code exist inside NOS/VE and you would likely need to be able to deadstart NOS/VE to be able to really use any of that.  There used to be cross tools that ran under NOS but I don't know of any existing copies.

**Reply by KC5VDJ** [reply]
I have to figure out where microcenter is, I know Denver has one, five dollar all day train/bus, and get a couple of the 0W2.

I wonder what the license terms for the antenna are?  I'd rather just have a SMA connector.  Such would make them even more useful.  Probably better than the connector cost.

At least I won't need new cases.  Do they sell them with the 40 pin attached at extra cost like before?  My soldering tools aren't even in this state at the moment and "helping hands" and panavise, either of which is necessary, given the size.

![IMG_20230906_142149733_HDR.jpg](https://github.com/kej715/DtCyber/assets/136194499/90549b88-85fd-492b-b4e6-532e0e7b23da)

Regardless, Pi-DI will target the single core version, as baseline.  Two years left on support, why toss perfectly usable controllers after October 2025?  Freebsd still have x10?  Control your house with the same thing that controlled nuclear power plants!   Need to see if fbsd arm6 added the missing devices on the pis.

**Comment by kej715** [comment]
I second William's comments, especially the bit about selecting the console
scheme via cyber.ini and having one console scheme that's equivalent to
what we have now.

The main reason I called attention to the 721 style console is that it was
a "simple" serial console interface and wouldn't require much, if any,
change to DtCyber itself. The existing tpmux.c module ought to support it,
or very little change would be needed. The work would be in implementing
the 721 emulator. As I've already implemented X.364, Pterm, Pterm Classic,
IBM 3270, and Tektronix 4014 emulators in Javascript for DtCyber,
implementing a 721 emulator probably would be a relatively light lift,
especially given that I have a Z80 emulator in Javascript too.

The main "trick" to implement would probably be customizing/modifying CTI
to support 721 mode. An alternative could be enhancing OSB to provide an
option for indicating that the console is a 721. Thinking aloud, perhaps
this could be done by designating some bits in the deadstart panel as bits
for selecting console type and 2-port mux port.

Downsides to relying on the 721 approach as the exclusive way of supporting
a "detached" console is that it would work only on mainframe models
supporting the 2-port mux (i.e., 800 series machines), and only most recent
versions of NOS would support it.

cheers,
Kevin

**Comment by KC5VDJ** [comment]
How does one get to the PostScript output?  Or was it in Diablo mode?  IIRC, it only did the two, or was there ibm or Epson too?  I only remember ps and Diablo 6xx, having done print utils under dos for the II Plus back in the the heady days before actual os postscript drivers .  Red book, yellow book, and blue book postscript here, no doubt.  It would seem rather pointless to buy a laserwriter just for Diablo emulation, especially at those prices, and the fact that a real Diablo would be cheaper and produce better characters (the Canon print engine was only 300 dpi).

![IMG_20230906_144857761_HDR.jpg](https://github.com/kej715/DtCyber/assets/136194499/cb7bca7e-114b-4c15-a8d4-d278c3828652)

Given the date on the guide, PC ignorance was quite evident still (maybe because of DOD "Desktop" 1 - 4  keeping crosstalk from going bankrupt).  By 1991, nobody used it.  Telix and Procomm were the only ones worth a damn at the time, and telix was shareware.  Hmmm, in fact, in '91, Procomm might have still been shareware, when did they sell out?l. PC ignorance was a large systems oldtimer willful frame of mind prior to the 90's largely, and a little into.  ("Why would anyone want to do more than one thing at a time or have windows" -- Don, he'd .h.. a brick seeing a cluster of dtcybers on a PC, cool guy though, knew Einstein, introduced me to Dr Feynman, i lost a bet, Don and Einstein actually did sail, Dr Feynman being a witness.  Ever see Goodfellas?  That's close to the Feynman accent.  Trippy van.  50% chance LIGO discovered a superstring recently dancing with a 4ish MS supermassive, he'd trip.  Times change).

**Reply by KC5VDJ** [reply]
And the 721 also does 3270 bsc?

**Comment by KC5VDJ** [comment]
If anyone can find a real 721, check out alt3 and alt5 pin functions in bcm2835 arm peripherals 6.2 (pp 102-103)...  these have two serial ports with rts/cts.  The other 232 pins can be emulated.  In theory a full 25 pin could be implemented, or used with ground into a couple of mmjs.  Oooh!  Dual session on a 420+...I'd love to get another vt420.  Fine terminal.  Helluva lot cheaper than a real vt525, no doubt (> $1000 on eBay for 525s, last checked)

**Comment by kej715** [comment]
No, the 721 did not do 3270 BSC.

**Reply by KC5VDJ** [reply]
It had a submodel on it.  Musta been like the 4014 and 4015, variation on a theme.  1991 cdcnet product descriptions.
